### DISCUSSION TITLE
MAUI and Uno Platform

### DISCUSSION THREAD

**Discussion by trungnt2910** [unoplatform/uno#8163] [Ideas]
As far as I know, Uno Platform provides support for WinUI 3 on WASM and on Linux GTK and Windows 7 (not covered by Xamarin) through Skia. The desktop and web targets performs decently and renders (nearly) perfect UI, through my development experience.

On the other hand, .NET MAUI seems to bring good support for mobile platforms, but does not support web browsers, earlier versions of Windows, and Linux.

Can the WinUI 3 support for .NET MAUI be ported to the Uno Platform, similar to [what Uno did with Xamarin and UWP](https://github.com/unoplatform/Uno.Xamarin.Forms.Platform), and bring MAUI development to desktop and web platforms too?

I understand that at the time this is written, MAUI is still in preview, but I'm just asking whether it's _possible_.

**Comment by francoistanguay** [comment]
While in theory we could do the same thing that we did with X.F Renderers for WebAssembly and use Uno to do MAUI => WinUI => WebAssembly and LInux, the feedback from our community showed that they'd rather have the fully featured WinUI API available to build apps versus the XF/MAUI abstraction.

Is there something that MAUI offers in terms of API or productivity that you feel is missing in WinUI/Uno?

**Reply by trungnt2910** [reply]
The API's perfect, but I'm seeking better performance for (low-end) mobile devices.

**Reply by francoistanguay** [reply]
Our metrics show differently , especially with all the work we've done in Uno 4.0 and the 4.1 we'll ship in the upcoming days.

If you do have an example of something that isnt as snappy, please try to share so we can profile/investigate. Performance is a top consideration.

**Reply by trungnt2910** [reply]
Well according to my dev experience a AOT-compiled Uno 4.0 application with a very basic UI (A `NavigationView` with one `TextBox` and a few `TextBlock`s) loads really slowly on Android (takes 8 seconds to load, on an Android device with 6GB of RAM and 8 cores).

I don't have similar problems for Desktop targets though.

**Reply by francoistanguay** [reply]
Do you mind me asking what the device model is or at least a comparable?

**Reply by trungnt2910** [reply]
It's a Samsung Galaxy A32.

On my old (low-end) phone the performance was much worse.
But debugging the app on WSA works great though.

**Reply by jeromelaban** [reply]
Are you building with AOT? Without it, loading time is very slow. Also, there are some devices where we suspect Xamarin.Android may select an incorrect ABIs. Could you provide an ADB logcat of the start of the app?

**Reply by trungnt2910** [reply]
> Are you building with AOT? Without it, loading time is very slow. Also, there are some devices where we suspect Xamarin.Android may select an incorrect ABIs. Could you provide an ADB logcat of the start of the app?

I am _certain_ that this is a release build with AOT enabled (and I have mentioned that I AOT-compiled it in the previous comment).

**Reply by jeromelaban** [reply]
Thanks, can you provide an ADB logcat ?

**Reply by trungnt2910** [reply]
> Thanks, can you provide an ADB logcat ?

Currently, I'm trying to get a logcat with:
```
adb logcat <my package name>:I *:S
```
to filter out unnecessary and potentially sensitive information, but the app seems to output nothing.

**Reply by jeromelaban** [reply]
If the app is running with AOT you'll get messages containing assembly names, when the runtime tries to find the files. Also, which version of Uno are you running?

**Reply by trungnt2910** [reply]
I'm using Uno.UI 4.0.13 (currently the latest release version of Uno).
But it's also slow with previous versions.

**Reply by jeromelaban** [reply]
Can you try with the latest 4.2 bits, we made significant changes.
